tests/integrations/android_sdk/test_verify_android_sdk.py

```diff
@@ -0,0 +1,281 @@
+    mock_command.shutil.unpack_archive.assert_not_called()
+
+    # The returned SDK has the expected root path.
+    assert sdk.root_path == root_path
+
+
+def test_user_provided_sdk(mock_command, tmp_path):
+    "If the user specifies a valid ANDROID_SDK_ROOT, it is used"
+    # Create `sdkmanager` and the license file.
+    root_path = tmp_path / "other_sdk"
+    tools_bin = root_path / "tools" / "bin"
+    tools_bin.mkdir(parents=True, mode=0o755)
+    (tools_bin / "sdkmanager").touch(mode=0o755)
+    licenses = root_path / "licenses"
+    licenses.mkdir(parents=True, mode=0o755)
+    (licenses / "android-sdk-license").touch()
+
+    # Set the environment to specify ANDROID_SDK_ROOT
+    mock_command.os.environ = {
+        'ANDROID_SDK_ROOT': str(root_path)
+    }
+
+    # Expect verify_android_sdk() to succeed
+    sdk = verify_android_sdk(mock_command)
+
+    # No calls to download, run or unpack anything.
+    mock_command.download_url.assert_not_called()
+    mock_command.subprocess.run.assert_not_called()
+    mock_command.subprocess.check_output.assert_not_called()
+    mock_command.shutil.unpack_archive.assert_not_called()
+
+    # The returned SDK has the expected root path.
+    assert str(sdk.root_path) == str(root_path)
+
+
+def test_invalid_user_provided_sdk(mock_command, tmp_path):
+    "If the user specifies an invalid ANDROID_SDK_ROOT, it is ignored"
+
+    # Create `sdkmanager` and the license file
+    # for the *briefcase* SDK.
+    root_path = tmp_path / ".briefcase" / "tools" / "android_sdk"
+    tools_bin = root_path / "tools" / "bin"
+    tools_bin.mkdir(parents=True, mode=0o755)
+    (tools_bin / "sdkmanager").touch(mode=0o755)
+    licenses = root_path / "licenses"
+    licenses.mkdir(parents=True, mode=0o755)
+    (licenses / "android-sdk-license").touch()
+
+    # Set the environment to specify an ANDROID_SDK_ROOT that doesn't exist
+    mock_command.os.environ = {
+        'ANDROID_SDK_ROOT': str(tmp_path / "other_sdk")
+    }
+
+    # Expect verify_android_sdk() to succeed
+    sdk = verify_android_sdk(mock_command)
+
+    # No calls to download, run or unpack anything.
+    mock_command.download_url.assert_not_called()
+    mock_command.subprocess.run.assert_not_called()
+    mock_command.subprocess.check_output.assert_not_called()
+    mock_command.shutil.unpack_archive.assert_not_called()
+
+    # The returned SDK has the expected root path.
+    assert sdk.root_path == root_path
+
+
+@pytest.mark.parametrize("host_os", ("ArbitraryNotWindows", "Windows"))
+def test_download_sdk(mock_command, tmp_path, host_os):
+    "If an SDK is not available, one will be downloaded"
+    root_path = tmp_path / ".briefcase" / "tools" / "android_sdk"
+
+    # Mock-out `host_os` so we only do our permission check on non-Windows.
+    mock_command.host_os = host_os
+
+    # The download will produce a cached file
+    cache_file = MagicMock()
+    cache_file.__str__.return_value = "/path/to/download.zip"
```

Comment:
I forgot again why you have to mock-out `__str__`. Can you leave a comment about why? Can you use a `Path` object?

Reply:
We need to be able to call "unlink" on the file, so we either need to create an actual file so it can be unlinked, or a mock object that we can verify. However, because it's mocking a Path object, it's being used as an argument to some of the process calls, where we call str() on the path because Python 3.5 Pathlib support isn't consistent; and str() on a mock doesn't yield a useful result.

We *could* use a Path object - but we'd then also need to create the physical file that could be deleted.